### PR TITLE
add clang_ci

### PR DESCRIPTION
**Pull request by Karsten1987**
as hinted at setup-ros https://github.com/ros-tooling/setup-ros/issues/276#issuecomment-827587265, there might be an easy way to setup a Clang build.

### CODE REVIEW

**Comment by Karsten1987** [comment]
@christophebedard it looks like the code snippet you've pasted led to a syntax error. I'd be very grateful if you could have a look at my PR here and tell me if you see anything obviously wrong with it.

File: .github/workflows/source.yml [hunk 1]
```diff
@@ -41,3 +41,30 @@ jobs:
...
           path: ros_ws/log
+
+  ci_source_clang:
+    name: Foxy source job
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+    steps:
+      - uses: ros-tooling/setup-ros@v0.1
+      - uses: ros-tooling/action-ros-ci@v0.1
```
Comment (christophebedard) [inline .github/workflows/source.yml:52]:
```suggestion
      - uses: ros-tooling/setup-ros@v0.2
      - uses: ros-tooling/action-ros-ci@v0.2
```

oops, you'll also need to bump this to be able to use `colcon-defaults`.

File: .github/workflows/source.yml [hunk 2]
```diff
@@ -41,3 +41,30 @@ jobs:
...
+          colcon-mixin-name: coverage-gcc
+          colcon-mixin-repository: https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml
+          colcon-defaults: |
+            {
+              "build": {
+                "mixin": [
+                    "clang-libcxx"
+                ]
+              }
+            }
```
Comment (christophebedard) [inline .github/workflows/source.yml:70]:
```suggestion
          colcon-mixin-repository: https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml
          colcon-defaults: |
            {
              "build": {
                "mixin": [
                    "coverage-gcc",
                    "clang-libcxx"
                ]
              }
            }
```

and move `coverage-gcc` to here

**Comment by christophebedard** [comment]
now the source jobs aren't showing up :thinking:

**Comment by Karsten1987** [comment]
hmm, yeah I just noticed the same.

**Comment by Karsten1987** [comment]
The linters also complain about "Unable to resolve action `ros-tooling/setup-ros@v0.2`, unable to find version `v0.2`"

**Comment by christophebedard** [comment]
> The linters also complain about "Unable to resolve action `ros-tooling/setup-ros@v0.2`, unable to find version `v0.2`"

argh it's `v0.1`, sorry!

**Comment by Karsten1987** [comment]
Thanks Christophe, that seems to run now - let's see if it also works ;-)

**Comment by Karsten1987** [comment]
so this seems to work, according to the output, Clang 10.0.0 is correctly used as a compiler. Looks like though that `libc++` is still different from whatever OSX has installed as it doesn't compile about `uint`.

**Comment by bmagyar** [comment]
What is the canonical OSX setup that should be supported?

**Comment by Karsten1987** [comment]
There's quite a bit of problems associated with OSX on GH action runners. Mainly, ROS2 on OSX requires SIP to be disabled, the GH action runners don't let us do this though (unfortunately). That means, you can't source your bash.setup correctly and things like DYLD_LIBRARY_PATH variables are not set correctly hindering a successful run of tests.
We have already an issue wo/ any succes: https://github.com/ros-tooling/action-ros-ci/issues/81

That's why I was hoping we could mimic some the OSX behavior by compiling with Clang. Seems as if `libc++` still deviates from OSX.

**Comment by bmagyar** [comment]
@Karsten1987 do you think there's value in finishing this? a master-only CI step would work fine

**Pull request closed without merging**